### PR TITLE
apt module :  take care about autoremove in upgrade function

### LINKED ISSUE #30516: apt module uses  apt-get --autoremove even if autoremove=no
##### ISSUE TYPE
 - Bug Report
##### COMPONENT NAME
apt
##### ANSIBLE VERSION
```
ansible 2.4.0.0 (detached HEAD 3c805f85bc) last updated 2017/09/18 18:23:57 (GMT +300)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/vashik/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/vashik/opt/ansible/lib/ansible
  executable location = /home/vashik/opt/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### OS / ENVIRONMENT

Debian Jessie

##### SUMMARY
If aptitude is missing apt module falls back to using apt-get but uses an option --autoremove that some versions of apt-get do not understood

##### STEPS TO REPRODUCE

Set up a minimal Debian Jessie system, it won't have aptitude installed.
Try to apply updates using apt module on it.

```bash
ansible -i target-host-name, target-host-name -m apt -sa 'upgrade=yes update_cache=yes autoclean=no autoremove=no' -vv
```

##### EXPECTED RESULTS

Apt module does not use --autoremove when autoremove=no
Apt module has default value no for autoremove
Apt module does not try to use --autoremove when apt-get does not support it 
##### ACTUAL RESULTS

```
target-host-name | FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "'/usr/bin/apt-get upgrade --with-new-pkgs --autoremove' failed: E: Command line option --autoremove is not understood\n",
    "rc": 100,
    "stdout": "",
    "stdout_lines": []
}
```

### PR DESCRIPTION
**Pull request by dj4ngo**
##### SUMMARY

Fixes #30516 
Take care about params autoremove in upgrade function. 

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

packaging/os/apt
##### ANSIBLE VERSION

```
ansible 2.5.0 (issue_30516 671de05a33) last updated 2017/09/22 13:51:54 (GMT +200)
```

### CODE REVIEW

**Comment by mscherer** [comment]
shipit

**Comment by dj4ngo** [comment]
cc @kbrebanov @elasticdog, I ping you as namespace maintener, to review this PR since mgwilliams seems not available.

**Comment by kbrebanov** [comment]
This would not change the functionality, but as a matter of consistency with the rest of the code, what would you think of passing the **autoremove** parameter when calling the **upgrade** function instead of looking up the value thru the module parameters?

Like when the **install** and **remove** functions are called.

**Comment by dj4ngo** [comment]
@kbrebanov I was thinking about it but I don't really agree with how args are managed...  
You're right, I'll push the change you are talking about for this PR

But I would like to talk a little bit about attributes management in this module. 
Since we get the ansible module m, we got two choices: 
 - to use it in each function to set the apt args to use (eg --autoremove), but this need to perform all if tests in each functions... 
 - create a function or code factorization to generate all cli args (with --autoremove in our case)
We could create a dictionary containing all the cli args for cmd (empty or with value) and pass it to each function to avoid having to modify args each-time we implement new functionnality. 
I'll propose a feature to implement this.

**Comment by kbrebanov** [comment]
@dj4ngo I think the **upgrade** call on line **984** may also need to be modified.

I agree with you that attributes management could probably be improved.

**Comment by dj4ngo** [comment]
@kbrebanov your right, I did a missmatch in my amend ....  I amend again my commit.

**Comment by kbrebanov** [comment]
shipit

**Comment by dj4ngo** [comment]
@mscherer Could you review this PR again ? I had to modify it since your shipit message ... Thanks

File: lib/ansible/modules/packaging/os/apt.py [hunk 1]
```diff
@@ -722,7 +722,13 @@ def cleanup(m, purge=False, force=False, operation=None,
 
 def upgrade(m, mode="yes", force=False, default_release=None,
             use_apt_get=False,
-            dpkg_options=expand_dpkg_options(DPKG_OPTIONS)):
+            dpkg_options=expand_dpkg_options(DPKG_OPTIONS), autoremove=None):
```
Comment (mscherer) [inline lib/ansible/modules/packaging/os/apt.py:725]: Shouldn't the default for autoremove be "False" rather than "None" ? (just nitpicking)

**Comment by mscherer** [comment]
shipit

**Comment by mscherer** [comment]
I suspect we are missing a few fixes. For example, what happen if there is no aptitude, but someone explictely say "autoremove=yes", the module should fail and/or warn the user, no ?

**Comment by abadger** [comment]
cherry-picked for 2.4.2beta1

**Pull request merged**